python/taichi/lang/matrix.py

```diff
@@ -269,21 +269,61 @@ def _get_slice(self, a, b):
             b = range(b.start or 0, b.stop or self.m, b.step or 1)
         return Matrix([[self(i, j) for j in b] for i in a])
 
+    def _cal_slice(self, index, dim):
+
+        start, stop, step = index.start or 0, index.stop or (
+            self.n if dim is 0 else self.m), index.step or 1
+        if isinstance(start, expr.Expr) and isinstance(start.ptr, int):
+            start = start.ptr
+        if isinstance(stop, expr.Expr) and isinstance(stop.ptr, int):
+            stop = stop.ptr
+        if isinstance(step, expr.Expr) and isinstance(step.ptr, int):
+            step = step.ptr
+        if isinstance(start, expr.Expr) or isinstance(
```

Comment:
IIUC you can have a small function to avoid these duplicated if checks: 
```
def helper(x, default):
    if x is None:
        return default
    if isinstance(x, expr.Expr):
        assert isinstance(x.ptr, int)
        return x.ptr
    return x
```
(feel free to edit, this might not be 100% what you want :P